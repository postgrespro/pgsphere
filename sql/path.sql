\set ECHO none
SELECT set_sphere_output_precision(8);
\set ECHO all

--
-- ellipse and path
--
  
-- negators , commutator @,&&

SELECT spath '{(280d, -9d),(280d, -8d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -9d),(280d,-12d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d,-11d),(280d,-12d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -9d),(280d, -8d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -9d),(280d,-12d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d,-11d),(280d,-12d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -9d),(280d, -8d)}'  !@  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -9d),(280d,-12d)}'  !@  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d,-11d),(280d,-12d)}'  !@  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -9d),(280d, -8d)}' !&&  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -9d),(280d,-12d)}' !&&  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d,-11d),(280d,-12d)}' !&&  sellipse '<{10d,5d},(280d,-20d),90d>'; 
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   ~  spath '{(280d, -9d),(280d, -8d)}'; 
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   ~  spath '{(280d, -9d),(280d,-12d)}'; 
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   ~  spath '{(280d,-11d),(280d,-12d)}'; 
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   &&  spath '{(280d, -9d),(280d, -8d)}';
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   &&  spath '{(280d, -9d),(280d,-12d)}';
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   &&  spath '{(280d,-11d),(280d,-12d)}';
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !~  spath '{(280d, -9d),(280d, -8d)}'; 
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !~  spath '{(280d, -9d),(280d,-12d)}'; 
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !~  spath '{(280d,-11d),(280d,-12d)}'; 
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !&&  spath '{(280d, -9d),(280d, -8d)}';
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !&&  spath '{(280d, -9d),(280d,-12d)}';
SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !&&  spath '{(280d,-11d),(280d,-12d)}';

-- path is a line , ellipse is point
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{0d,0d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{0d,0d},(280d, -8d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{0d,0d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{0d,0d},(280d, -8d),90d>';
-- path is a line , ellipse is circle
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,5d},(280d, -8d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,5d},(280d, -8d),90d>';
-- path is a line , ellipse is path
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,0d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,0d},(280d, -8d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,0d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,0d},(280d, -8d),90d>';
-- path is a line , ellipse is a real ellipse
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{10d,5d},(280d, -8d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{10d,5d},(280d, -8d),90d>';
-- create path as aggregate
SELECT spath(data.p) FROM ( SELECT spoint '(0,1)' as p UNION ALL SELECT spoint '(1,1)' UNION ALL SELECT '(1,0)' ) AS data ;

--
-- checking path other operators
--
  
\set poly  'spoly \'{(0.1,0),(0.2,0),(0.2,0.1),(0.3,0.1),(0.3,-0.1),(0.4,-0.1),(0.5,0.1),(0.4,0.2),(0.1,0.2)}\''
\set path1 'spath \'{(0.1,0),(0.2,0),(0.2,0.1),(0.3,0.1),(0.3,-0.1),(0.4,-0.1),(0.5,0.1),(0.4,0.2),(0.1,0.2)}\''
\set path2 'spath \'{(0,0),(1,0),(2,0),(3,0)}\''
\set path3 'spath \'{(0,0),(0,1),(0,1.5)}\''

SELECT @-@ spath '{(0,0),(1,0),(2,0),(3,0)}';
SELECT :path1 =  :path2;
SELECT :path1 =  :path1;
SELECT :path1 <> :path2;
SELECT :path1 <> :path1;
SELECT :poly  && :path1;
SELECT :path1 && :poly ;
SELECT :path1 @  :poly ;
SELECT :path2 @  :poly ;
SELECT :path1 && :path1;
SELECT :path1 && :path1;
SELECT :poly  && :path2;
SELECT :path2 && :poly ;
SELECT :path2 && :path1;
SELECT :poly  && :path3;
SELECT :path3 && :poly ;
SELECT :path3 && :path1;
SELECT :path3 && :path2;
SELECT :path1 @  scircle '<(0,1),1>';
SELECT :path3 @  scircle '<(0,1),1>';
SELECT :path3 @  scircle '<(0,1),0.7>';
SELECT :path1 && scircle '<(0,1),1>';  
SELECT :path3 && scircle '<(0,1),1>';  
SELECT :path3 && scircle '<(0,1),0.7>';
SELECT :path3 && scircle '<(0,-1),0.7>';
SELECT :path3 @  scircle '<(0,-1),0.7>';
SELECT :path3 && sline ( spoint '(0,-1)', spoint '(0,1)' );
SELECT :path3 && sline ( spoint '(-1,0)', spoint '(1,0)' );
SELECT :path3 && sline ( spoint '(-1,0)', spoint '(-0.3,0)' );
SELECT spath '{(0.11,0.15),(0.12,0.15),(0.13,0.15)}' @ :poly; 

-- create path
SELECT spath(data.p) FROM ( SELECT spoint '(0,1)' as p UNION ALL SELECT spoint '(1,1)' UNION ALL SELECT '(1,0)' ) AS data ;

SELECT set_sphere_output( 'DEG' );

-- test stored data
SELECT spoint(p,2) FROM spheretmp6 WHERE id=2;

SELECT set_sphere_output( 'RAD' );

-- get n-th point and array representation path points tests
SELECT spoint( spath '{(0, 0),(1, 1)}', 1 );
SELECT spoint( spath '{(0, 0),(1, 1)}', 2 );
SELECT spath_as_array( spath '{(0, 0),(1, 1)}');
