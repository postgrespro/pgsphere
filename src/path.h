#ifndef __PGS_PATH_H__
#define __PGS_PATH_H__

#include "polygon.h"

/* Path declarations */


/*
 * The definition of spherical path using a list of
 * spherical points.
 */
typedef struct
{
	char	vl_len_[4];	/* total size in bytes */
	int32	npts;		/* count of points */
	SPoint	p[1];		/* variable length array of SPoints */
} SPATH;



/* Path and circle */
#define PGS_CIRCLE_PATH_AVOID 0		/* circle avoids path */
#define PGS_CIRCLE_CONT_PATH  1		/* circle contains path */
#define PGS_CIRCLE_PATH_OVER  2		/* circle overlaps path */

/* Path and polygon */
#define PGS_POLY_PATH_AVOID   0		/* polygon avoids path */
#define PGS_POLY_CONT_PATH	  1		/* polygon contains path */
#define PGS_POLY_PATH_OVER	  2		/* polygon and path overlap */

/* Path and ellipse */
#define PGS_ELLIPSE_PATH_AVOID 0	/* ellipse avoids path */
#define PGS_ELLIPSE_CONT_PATH  1	/* ellipse contains path */
#define PGS_ELLIPSE_PATH_OVER  2	/* ellipse overlaps path */


#define PG_GETARG_SPATH(arg) \
	( (SPATH *) DatumGetPointer(PG_DETOAST_DATUM(PG_GETARG_DATUM(arg))) )

/*
 * Checks whether two paths are equal.
 */
bool	spath_eq(const SPATH *p1, const SPATH *p2);

/*
 * Checks whether a path contains a point.
 */
bool	spath_cont_point(const SPATH *path, const SPoint *sp);

/*
 * Returns the i-th line segment of a path.
 *
 * sl   - pointer to line segment
 * path - pointer to path
 * i    - number of
 *
 * Returns the pointer to the line segment or NULL if fails.
 */
bool	spath_segment(SLine *sl, const SPATH *path, int32 i);

/*
 * Input function of path.
 */
Datum	spherepath_in(PG_FUNCTION_ARGS);

/*
 * Returns the n-th point of a path.
 *
 * See spherepath_get_point(PG_FUNCTION_ARGS)
 */
Datum	spherepath_get_point(PG_FUNCTION_ARGS);

/*
 * This function interpolates between points of path. Returns the
 * n-th point of a path where n is a float.
 *
 * See spherepath_point(PG_FUNCTION_ARGS)
 */
Datum	spherepath_point(PG_FUNCTION_ARGS);

/*
 * Checks whether two paths are equal.
 */
Datum	spherepath_equal(PG_FUNCTION_ARGS);

/*
 * Checks whether two paths aren't equal.
 */
Datum	spherepath_equal_neg(PG_FUNCTION_ARGS);

/*
 * Returns the length of a path.
 */
Datum	spherepath_length(PG_FUNCTION_ARGS);

/*
 * Returns the number of points in a path.
 */
Datum	spherepath_npts(PG_FUNCTION_ARGS);

/*
 * Changes the direction of a path.
 */
Datum	spherepath_swap(PG_FUNCTION_ARGS);

/*
 * Checks whether a path contains a point.
 */
Datum	spherepath_cont_point(PG_FUNCTION_ARGS);

/*
 * Checks whether a path doesn't contain a point.
 */
Datum	spherepath_cont_point_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a path contains a point.
 */
Datum	spherepath_cont_point_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a path doesn't contain a point.
 */
Datum	spherepath_cont_point_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a path and a line overlap.
 */
Datum	spherepath_overlap_line(PG_FUNCTION_ARGS);

/*
 * Checks whether a path and a line don't overlap.
 */
Datum	spherepath_overlap_line_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether path and line overlap.
 */
Datum	spherepath_overlap_line_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a path and a line don't overlap.
 */
Datum	spherepath_overlap_line_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a path.
 */
Datum	spherecircle_cont_path(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a path.
 */
Datum	spherecircle_cont_path_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a path.
 */
Datum	spherecircle_cont_path_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a path.
 */
Datum	spherecircle_cont_path_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle and a path overlap.
 */
Datum	spherecircle_overlap_path(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle and a path don't overlap.
 */
Datum	spherecircle_overlap_path_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle and a path overlap.
 */
Datum	spherecircle_overlap_path_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle and a path don't overlap.
 */
Datum	spherecircle_overlap_path_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon contains a path.
 */
Datum	spherepoly_cont_path(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon doesn't contain a path.
 */
Datum	spherepoly_cont_path_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon contains path.
 */
Datum	spherepoly_cont_path_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon doesn't contain a path.
 */
Datum	spherepoly_cont_path_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon and a path overlap.
 */
Datum	spherepoly_overlap_path(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon and a path don't overlap.
 */
Datum	spherepoly_overlap_path_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon and a path overlap.
 */
Datum	spherepoly_overlap_path_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a polygon and a path don't overlap.
 */
Datum	spherepoly_overlap_path_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether two paths overlap.
 */
Datum	spherepath_overlap_path(PG_FUNCTION_ARGS);

/*
 * Checks whether two paths don't overlap.
 */
Datum	spherepath_overlap_path_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse contains path.
 */
Datum	sphereellipse_cont_path(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse doesn't contain a path.
 */
Datum	sphereellipse_cont_path_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse contains a path.
 */
Datum	sphereellipse_cont_path_com(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse doesn't contain a path.
 */
Datum	sphereellipse_cont_path_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse and a path overlap.
 */
Datum	sphereellipse_overlap_path(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse and a path don't overlap.
 */
Datum	sphereellipse_overlap_path_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse and a path overlap.
 */
Datum	sphereellipse_overlap_path_com(PG_FUNCTION_ARGS);

/*
 * Checks whether an ellipse and a path don't overlap.
 */
Datum	sphereellipse_overlap_path_com_neg(PG_FUNCTION_ARGS);

/*
 * Performs an Euler transformation on a path.
 */
Datum	spheretrans_path(PG_FUNCTION_ARGS);

/*
 * Performs an inverse Euler transformation on a path.
 */
Datum	spheretrans_path_inverse(PG_FUNCTION_ARGS);

/*
 * State transition function for aggregate function spath(spoint). Never
 * call this function outside an aggregate function!
 *
 * Adds a point to a path.
 */
Datum	spherepath_add_point(PG_FUNCTION_ARGS);

/*
 * Finalize function for adding spoints to a path.
 */
Datum	spherepath_add_points_finalize(PG_FUNCTION_ARGS);

#endif
