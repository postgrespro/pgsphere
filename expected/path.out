\set ECHO none
 set_sphere_output_precision 
-----------------------------
 SET 8
(1 row)

--
-- ellipse and path
--
  
-- negators , commutator @,&&
SELECT spath '{(280d, -9d),(280d, -8d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -9d),(280d,-12d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d,-11d),(280d,-12d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d, -9d),(280d, -8d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -9d),(280d,-12d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d,-11d),(280d,-12d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d, -9d),(280d, -8d)}'  !@  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d, -9d),(280d,-12d)}'  !@  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d,-11d),(280d,-12d)}'  !@  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -9d),(280d, -8d)}' !&&  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d, -9d),(280d,-12d)}' !&&  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d,-11d),(280d,-12d)}' !&&  sellipse '<{10d,5d},(280d,-20d),90d>'; 
 ?column? 
----------
 f
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   ~  spath '{(280d, -9d),(280d, -8d)}'; 
 ?column? 
----------
 f
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   ~  spath '{(280d, -9d),(280d,-12d)}'; 
 ?column? 
----------
 f
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   ~  spath '{(280d,-11d),(280d,-12d)}'; 
 ?column? 
----------
 t
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   &&  spath '{(280d, -9d),(280d, -8d)}';
 ?column? 
----------
 f
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   &&  spath '{(280d, -9d),(280d,-12d)}';
 ?column? 
----------
 t
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'   &&  spath '{(280d,-11d),(280d,-12d)}';
 ?column? 
----------
 t
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !~  spath '{(280d, -9d),(280d, -8d)}'; 
 ?column? 
----------
 t
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !~  spath '{(280d, -9d),(280d,-12d)}'; 
 ?column? 
----------
 t
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !~  spath '{(280d,-11d),(280d,-12d)}'; 
 ?column? 
----------
 f
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !&&  spath '{(280d, -9d),(280d, -8d)}';
 ?column? 
----------
 t
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !&&  spath '{(280d, -9d),(280d,-12d)}';
 ?column? 
----------
 f
(1 row)

SELECT sellipse '<{10d,5d},(280d,-20d),90d>'  !&&  spath '{(280d,-11d),(280d,-12d)}';
 ?column? 
----------
 f
(1 row)

-- path is a line , ellipse is point
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{0d,0d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{0d,0d},(280d, -8d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{0d,0d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{0d,0d},(280d, -8d),90d>';
 ?column? 
----------
 t
(1 row)

-- path is a line , ellipse is circle
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,5d},(280d, -8d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,5d},(280d, -8d),90d>';
 ?column? 
----------
 t
(1 row)

-- path is a line , ellipse is path
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,0d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{5d,0d},(280d, -8d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,0d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{5d,0d},(280d, -8d),90d>';
 ?column? 
----------
 t
(1 row)

-- path is a line , ellipse is a real ellipse
SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'   @  sellipse '<{10d,5d},(280d, -8d),90d>';
 ?column? 
----------
 t
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{10d,5d},(280d,-20d),90d>';
 ?column? 
----------
 f
(1 row)

SELECT spath '{(280d, -8d),(280d, -9d)}'  &&  sellipse '<{10d,5d},(280d, -8d),90d>';
 ?column? 
----------
 t
(1 row)

-- create path as aggregate
SELECT spath(data.p) FROM ( SELECT spoint '(0,1)' as p UNION ALL SELECT spoint '(1,1)' UNION ALL SELECT '(1,0)' ) AS data ;
           spath           
---------------------------
 {(0 , 1),(1 , 1),(1 , 0)}
(1 row)

--
-- checking path other operators
--
  
\set poly  'spoly \'{(0.1,0),(0.2,0),(0.2,0.1),(0.3,0.1),(0.3,-0.1),(0.4,-0.1),(0.5,0.1),(0.4,0.2),(0.1,0.2)}\''
\set path1 'spath \'{(0.1,0),(0.2,0),(0.2,0.1),(0.3,0.1),(0.3,-0.1),(0.4,-0.1),(0.5,0.1),(0.4,0.2),(0.1,0.2)}\''
\set path2 'spath \'{(0,0),(1,0),(2,0),(3,0)}\''
\set path3 'spath \'{(0,0),(0,1),(0,1.5)}\''
SELECT @-@ spath '{(0,0),(1,0),(2,0),(3,0)}';
 ?column? 
----------
        3
(1 row)

SELECT :path1 =  :path2;
 ?column? 
----------
 f
(1 row)

SELECT :path1 =  :path1;
 ?column? 
----------
 t
(1 row)

SELECT :path1 <> :path2;
 ?column? 
----------
 t
(1 row)

SELECT :path1 <> :path1;
 ?column? 
----------
 f
(1 row)

SELECT :poly  && :path1;
 ?column? 
----------
 t
(1 row)

SELECT :path1 && :poly ;
 ?column? 
----------
 t
(1 row)

SELECT :path1 @  :poly ;
 ?column? 
----------
 t
(1 row)

SELECT :path2 @  :poly ;
 ?column? 
----------
 f
(1 row)

SELECT :path1 && :path1;
 ?column? 
----------
 t
(1 row)

SELECT :path1 && :path1;
 ?column? 
----------
 t
(1 row)

SELECT :poly  && :path2;
 ?column? 
----------
 t
(1 row)

SELECT :path2 && :poly ;
 ?column? 
----------
 t
(1 row)

SELECT :path2 && :path1;
 ?column? 
----------
 t
(1 row)

SELECT :poly  && :path3;
 ?column? 
----------
 f
(1 row)

SELECT :path3 && :poly ;
 ?column? 
----------
 f
(1 row)

SELECT :path3 && :path1;
 ?column? 
----------
 f
(1 row)

SELECT :path3 && :path2;
 ?column? 
----------
 t
(1 row)

SELECT :path1 @  scircle '<(0,1),1>';
 ?column? 
----------
 f
(1 row)

SELECT :path3 @  scircle '<(0,1),1>';
 ?column? 
----------
 t
(1 row)

SELECT :path3 @  scircle '<(0,1),0.7>';
 ?column? 
----------
 f
(1 row)

SELECT :path1 && scircle '<(0,1),1>';  
 ?column? 
----------
 t
(1 row)

SELECT :path3 && scircle '<(0,1),1>';  
 ?column? 
----------
 t
(1 row)

SELECT :path3 && scircle '<(0,1),0.7>';
 ?column? 
----------
 t
(1 row)

SELECT :path3 && scircle '<(0,-1),0.7>';
 ?column? 
----------
 f
(1 row)

SELECT :path3 @  scircle '<(0,-1),0.7>';
 ?column? 
----------
 f
(1 row)

SELECT :path3 && sline ( spoint '(0,-1)', spoint '(0,1)' );
 ?column? 
----------
 t
(1 row)

SELECT :path3 && sline ( spoint '(-1,0)', spoint '(1,0)' );
 ?column? 
----------
 t
(1 row)

SELECT :path3 && sline ( spoint '(-1,0)', spoint '(-0.3,0)' );
 ?column? 
----------
 f
(1 row)

SELECT spath '{(0.11,0.15),(0.12,0.15),(0.13,0.15)}' @ :poly; 
 ?column? 
----------
 t
(1 row)

-- create path
SELECT spath(data.p) FROM ( SELECT spoint '(0,1)' as p UNION ALL SELECT spoint '(1,1)' UNION ALL SELECT '(1,0)' ) AS data ;
           spath           
---------------------------
 {(0 , 1),(1 , 1),(1 , 0)}
(1 row)

SELECT set_sphere_output( 'DEG' );
 set_sphere_output 
-------------------
 SET DEG
(1 row)

-- test stored data
SELECT spoint(p,2) FROM spheretmp6 WHERE id=2;
   spoint   
------------
 (1d , -5d)
(1 row)

